pkg/controller/nstemplateset/nstemplateset_controller_test.go

```diff
@@ -0,0 +1,349 @@
+	// Issue: https://github.com/kubernetes-sigs/controller-runtime/issues/524
+
+	// uncomment below code block when issue fixed
+	// labels := map[string]string{"owner": username, "type": typeName}
+	// opts := client.MatchingLabels(labels)
+	// namespaceList := &corev1.NamespaceList{}
+	// err := cl.List(context.TODO(), opts, namespaceList)
+	// require.NoError(t, err)
+	// require.NotNil(t, namespaceList)
+	// require.Equal(t, 1, len(namespaceList.Items))
+	// return &namespaceList.Items[0]
+
+	// remove below code block when issue fixed
+	namespace := &corev1.Namespace{}
+	nsName := fmt.Sprintf("%s-%s", username, typeName)
+	err := cl.Get(context.TODO(), types.NamespacedName{Name: nsName}, namespace)
+	require.NoError(t, err)
+	require.Equal(t, namespace.Labels["owner"], username)
+	require.Equal(t, namespace.Labels["type"], typeName)
+	return namespace
+}
+
+func checkInnerResources(t *testing.T, client *test.FakeClient, nsName string) {
+	t.Helper()
+
+	roleBinding := &authv1.RoleBinding{}
+	err := client.Get(context.TODO(), types.NamespacedName{Name: "user-edit", Namespace: nsName}, roleBinding)
+	require.NoError(t, err)
+}
+
+func checkStatus(t *testing.T, client *test.FakeClient, username string, wantStatus corev1.ConditionStatus, wantReason string) {
+	t.Helper()
+
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{}
+	err := client.Get(context.TODO(), types.NamespacedName{Name: username, Namespace: namespaceName}, nsTmplSet)
+	require.NoError(t, err)
+	readyCond, found := condition.FindConditionByType(nsTmplSet.Status.Conditions, toolchainv1alpha1.ConditionReady)
+	assert.True(t, found)
+	assert.Equal(t, wantStatus, readyCond.Status)
+	assert.Equal(t, wantReason, readyCond.Reason)
+}
+
+func newNSTmplSet(userName string) *toolchainv1alpha1.NSTemplateSet {
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      userName,
+			Namespace: namespaceName,
+		},
+		Spec: toolchainv1alpha1.NSTemplateSetSpec{
+			TierName: "basic",
+			Namespaces: []toolchainv1alpha1.NSTemplateSetNamespace{
+				{Type: "dev", Revision: "abcde11", Template: ""},
+				{Type: "code", Revision: "abcde21", Template: ""},
+			},
+		},
+	}
+	return nsTmplSet
+}
+
+func prepareReconcile(t *testing.T, username string, initObjs ...runtime.Object) (*ReconcileNSTemplateSet, reconcile.Request, *test.FakeClient) {
+	s := scheme.Scheme
+	err := apis.AddToScheme(s)
+	require.NoError(t, err)
+	codecFactory := serializer.NewCodecFactory(s)
+	decoder := codecFactory.UniversalDeserializer()
+
+	fakeClient := test.NewFakeClient(t, initObjs...)
+	r := &ReconcileNSTemplateSet{
+		client:             fakeClient,
+		scheme:             s,
+		getTemplateContent: wrapTetTemplateContent(decoder),
+	}
+	return r, newReconcileRequest(username), fakeClient
+}
+
+func newReconcileRequest(name string) reconcile.Request {
+	return reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      name,
+			Namespace: namespaceName,
```

Comment:
I've noticed that you don't have any test case where namespace is not, so you cannot verify this part https://github.com/codeready-toolchain/member-operator/pull/53/files#diff-cffea82f7fdc7427b8c9f0d0fdd55aedR94

Reply:
The link is broken, I am not getting your point.  Can you elaborate plz.

Reply:
that you don't verify the case when the reconcile method is triggered with a request that doesn't have namespace parameter set - as a reaction on creation/changes of namespaces.
https://github.com/codeready-toolchain/member-operator/pull/53/files#diff-cffea82f7fdc7427b8c9f0d0fdd55aedR94-R100
I'm talking about this part in the reconcile:
```go
if namespace == "" {
	namespace, err = k8sutil.GetWatchNamespace()
```
```

Reply:
Done https://github.com/codeready-toolchain/member-operator/pull/53/commits/424f4d10966946a30f1e8cc75256c27da52df09c